src/lib/reasoners/arith.ml

```diff
@@ -134,39 +134,47 @@ module Shostak
     | Some p -> p
     | _ -> P.create [Q.one, r] Q.zero (X.type_info r)
 
-  (* t1 % t2 = md  <->
-     c1. 0 <= md ;
-     c2. md < t2 ;
-     c3. exists k. t1 = t2 * k + t ;
-     c4. t2 <> 0 (already checked) *)
-  let mk_modulo md t1 t2 p2 ctx =
-    let zero = E.int "0" in
-    let c1 = E.mk_builtin ~is_pos:true Symbols.LE [zero; md] in
-    let c2 =
-      match P.is_const p2 with
-      | Some n2 ->
-        let an2 = Q.abs n2 in
-        assert (Q.is_int an2);
-        let t2 = E.int (Q.to_string an2) in
-        E.mk_builtin ~is_pos:true Symbols.LT [md; t2]
-      | None ->
-        E.mk_builtin ~is_pos:true Symbols.LT [md; t2]
-    in
-    let k  = E.fresh_name Ty.Tint in
-    let t3 = E.mk_term (Sy.Op Sy.Mult) [t2;k] Ty.Tint in
-    let t3 = E.mk_term (Sy.Op Sy.Plus) [t3;md] Ty.Tint in
-    let c3 = E.mk_eq ~iff:false t1 t3 in
-    c3 :: c2 :: c1 :: ctx
-
-  let mk_euc_division p p2 t1 t2 ctx =
-    match P.to_list p2 with
-    | [], coef_p2 ->
-      let md = E.mk_term (Sy.Op Sy.Modulo) [t1;t2] Ty.Tint in
-      let r, ctx' = X.make md in
-      let rp =
-        P.mult_const (Q.div Q.one coef_p2) (embed r) in
-      P.sub p rp, ctx' @ ctx
-    | _ -> assert false
+  (* Compute [t1 % t2] as [t1 - |t2| * (t1 / |t2|)].
+
+     Requires: [t1], [t2] have type [Tint]
+     Requires: [p1] is the representative for [t1]
+     Requires: [p2] is the representative for [t2]
+     Requires: [p2] is a non-zero constant *)
+  let mk_modulo t1 p1 p2 ctx =
+    match P.is_const p2 with
+    | Some n2 ->
+      let a2 = Q.to_z n2 |> Z.abs in
+      assert (Z.sign a2 > 0);
+
+      (* NB: [t1 % t2] = [t1 % |t2|] *)
```

Comment:
Here the equality `=` denotes the usual equality relation between integers and this equation is consequence of our choice of representative term for the equivalence classes.

Reply:
> This follows from the definition of `t1 % t2` as the remainder of euclidean division.

It's exactly what I said ;)